p2p/chain.py

```diff
@@ -437,20 +486,50 @@ def __init__(self,
 
         return dict((part.unique_key, part.part) for part in parts)
 
+    def _get_throughput_stats(self, peer : BasePeer, body_part: BlockPartEnum) -> ThroughputTracker:
+        return self._throughput_stats[body_part][peer]
+
     def _request_block_parts(
             self,
             target_td: int,
             headers: List[BlockHeader],
-            request_func: Callable[[ETHPeer, List[BlockHeader]], None]) -> int:
+            request_func: Callable[[ETHPeer, List[BlockHeader]], None],
+            block_part: BlockPartEnum,
+    ) -> int:
         peers = self.peer_pool.get_peers(target_td)
         if not peers:
             raise NoEligiblePeers()
-        length = math.ceil(len(headers) / len(peers))
-        batches = list(partition_all(length, headers))
+        batches = self._select_headers_by_peer(headers, peers, block_part)
         for peer, batch in zip(peers, batches):
+            self._get_throughput_stats(peer, block_part).begin_work()
```

Comment:
Should this handle the case where work was already begun but never completed?  Not entirely sure but it seems like a peer could disconnect and leave us in a *middle* state at which point this will fail everytime we try to request things from a given peer.

Reply:
Added a `deregister_peer` to remove the stats when a peer logs off. This also has the nice effect of limiting the total size of the stats we track.